Plugins/SpatialGDK/SpatialGDKEditorToolbar/Source/SpatialGDKEditorToolbar/Private/InteropCodeGenerator/TypeBindingGenerator.cpp

```diff
@@ -149,6 +151,87 @@ FString PropertyToWorkerSDKType(UProperty* Property, bool bIsRPCProperty)
 	return DataType;
 }
 
+// Generate definition in header file for Blueprint-defined Structs & Enums
+void GenerateBlueprintStructEnumDefinition(FCodeWriter& Writer, UProperty* Property, TSet<FString>& GeneratedTypes)
+{
+	if (GeneratedTypes.Contains(Property->GetCPPType()))
+	{
+		return;
+	}
+
+	if (UArrayProperty *ArrProp = Cast<UArrayProperty>(Property))
+	{
+		GenerateBlueprintStructEnumDefinition(Writer, ArrProp->Inner, GeneratedTypes);
+	}
+
+	if (Property->IsA<UStructProperty>())
+	{
+		UStruct* Struct = Cast<UStructProperty>(Property)->Struct;
+		if (!Struct->IsA<UUserDefinedStruct>())
+		{
+			return;
+		}
+
+		GeneratedTypes.Add(Property->GetCPPType());
+
+		// Recurse children post-order
+		for (TFieldIterator<UProperty> PropIt(Struct); PropIt; ++PropIt)
+		{
+			UProperty* Prop = *PropIt;
+			GenerateBlueprintStructEnumDefinition(Writer, Prop, GeneratedTypes);
+		}
+
+		Writer.PrintNewLine();
+		Writer.Printf("struct %s", *Property->GetCPPType());
+		Writer.Printf("{").Indent();
+		for (TFieldIterator<UProperty> PropIt(Struct); PropIt; ++PropIt)
+		{
+			UProperty* Prop = *PropIt;
+			FStringOutputDevice PropertyText;
+			Prop->ExportCppDeclaration(PropertyText, EExportedDeclaration::Local, nullptr);
+			Writer.Printf("%s;", *PropertyText);
+		}
+		Writer.Outdent().Printf("};");
+	}
+	else
+	{
+		// Enums could be represented as either UByteProperty or UEnumProperty
+		UEnum* Enum = nullptr;
+		if (Property->IsA<UByteProperty>())
+		{
+			Enum = Cast<UByteProperty>(Property)->Enum;
+		}
+		else if (Property->IsA<UEnumProperty>())
+		{
+			Enum = Cast<UEnumProperty>(Property)->GetEnum();
+		}
+
+		if (Enum == nullptr)
+		{
+			return;
+		}
+
+		if (!Enum->IsA<UUserDefinedEnum>())
+		{
+			return;
+		}
+
+		GeneratedTypes.Add(Property->GetCPPType());
+
+		Writer.PrintNewLine();
+		Writer.Printf("enum %s", *Enum->GetName());
+		Writer.Printf("{").Indent();
+
+		int ind = 0;
+		for (; ind < Enum->NumEnums() - 1; ind++)
```

Comment:
Can you put the `ind` inside the `for`?

Reply:
We use it for the final count; seems cleaner than `Enum->NumEnums()-1` again

Reply:
I really dislike how that for loop looks. I think when we use it again outside of the loop, doing `Enum->NumEnums()-1` is fine, it's explicit.

Reply:
Thoughts @improbable-valentyn ?

Reply:
Fixed.